website/blog/2023-02-14-pgo.md

```diff
@@ -0,0 +1,101 @@
+
+PGO allows data to be collected based on application scenarios in a production environment, allowing the optimiser to optimise speed for hot code paths and size for cold code paths, producing faster and smaller code for applications.
+
+rustc supports PGO, with data collection built into the binaries and then perf data collected during runtime in preparation for final compilation optimisation. The implementation relies entirely on LLVM.
+
+## Workflow
+
+Generating a PGO-optimized program involves following a workflow with four steps:
+
+- Compile the program with instrumentation enabled.
+- Run the instrumented program which generates a `profraw` file.
+- Convert the `.profraw` file into a `.profdata` file using LLVM's `llvm-profdata` tool.
+- Compile the program again, this time making use of the profiling data.
+
+## Preparation
+
+The data collected during the run will eventually need to be converted with `llvm-profdata`, which can be provided by installing the `llvm-tools-preview` component via `rustup`, or consider using recent versions of LLVM and the program provided by Clang.
+
+```bash
+rustup component add llvm-tools-preview
+```
+
+After installation `llvm-profdata` may need to be added to the `PATH` at the following path.
+
+```bash
+~/.rustup/toolchains/<toolchain>/lib/rustlib/<target-triple>/bin/
+```
+
+## Step-By-Step
+
+Rather than choosing a specific production environment workload, Databend's SQL logic tests are used here as a demonstration. It may not be positive in performance, but it can help us to understand how the process is carried out.
+
+> The caveat, however, is that the sample of data fed to the program during the profiling stage must be statistically representative of the typical usage scenarios; otherwise, profile-guided feedback has the potential to harm the overall performance of the final build instead of improving it.
+
+1. Make sure there is no left-over profiling data from previous runs.
+
+  ```bash
+  rm -rf /tmp/pgo-data
+  ```
+
+2. Build the instrumented binaries (with `release` profile), using the `RUSTFLAGS` environment variable in order to pass the PGO compiler flags to the compilation of all crates in the program.
+
+  ```bash
+  RUSTFLAGS="-Cprofile-generate=/tmp/pgo-data" \
+      cargo build --release --target=x86_64-unknown-linux-gnu
+  ```
+
+3. Instrumented binaries were run with some typical workload and we strongly recommend using workload that is statistically representative of the real scenario. This example runs SQL logic tests for reference only.
+  - Start a stand-alone Databend via a script, or a Databend cluster, given that production environments are more likely to run in cluster mode.
```

Comment:
given that production environments are more likely to run in cluster mode.-> Note that a production environment is more likely to run in cluster mode.